registry/common/database.py

```diff
@@ -0,0 +1,153 @@
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+import logging
+import threading
+import os
+import pymssql
+
+# TODO: refactor common utils in registry folder
+# This is a copy of database.py from sql-registry folder 
+# And add a `update` function to execute insert and set commands
+
+providers = []
+
+
+class DbConnection(ABC):
+    @abstractmethod
+    def query(self, sql: str, *args, **kwargs) -> list[dict]:
+        pass
+
+
+def quote(id):
+    if isinstance(id, str):
+        return f"'{id}'"
+    else:
+        return ",".join([f"'{i}'" for i in id])
+
+
+def parse_conn_str(s: str) -> dict:
+    """
+    TODO: Not a sound and safe implementation, but useful enough in this case
+    as the connection string is provided by users themselves.
+    """
+    parts = dict([p.strip().split("=", 1)
+                 for p in s.split(";") if len(p.strip()) > 0])
+    server = parts["Server"].split(":")[1].split(",")[0]
+    return {
+        "host": server,
+        "database": parts["Initial Catalog"],
+        "user": parts["User ID"],
+        "password": parts["Password"],
+        # "charset": "utf-8",   ## For unknown reason this causes connection failure
+    }
+
+
+class MssqlConnection(DbConnection):
+    @staticmethod
+    def connect(autocommit=True):
+        conn_str = os.environ["CONNECTION_STR"]
+        if "Server=" not in conn_str:
+            logging.debug(
+                "`CONNECTION_STR` is not in ADO connection string format")
+            return None
+        params = parse_conn_str(conn_str)
+        if not autocommit:
+            params["autocommit"] = False
+        return MssqlConnection(params)
+
+    def __init__(self, params):
+        self.params = params
+        self.make_connection()
+        self.mutex = threading.Lock()
+
+    def make_connection(self):
+        self.conn = pymssql.connect(**self.params)
+
+    def query(self, sql: str, *args, **kwargs) -> list[dict]:
+        """
+        Make SQL query and return result
+        """
+        logging.debug(f"SQL: `{sql}`")
+        # NOTE: Only one cursor is allowed at the same time
+        retry = 0
+        while True:
+            try:
+                with self.mutex:
+                    c = self.conn.cursor(as_dict=True)
+                    c.execute(sql, *args, **kwargs)
+                    return c.fetchall()
+            except pymssql.OperationalError:
+                logging.warning("Database error, retrying...")
+                # Reconnect
+                self.make_connection()
+                retry += 1
+                if retry >= 3:
+                    # Stop retrying
+                    raise
+                pass
+
+    def update(self, sql: str, *args, **kwargs):
```

Comment:
It seems werid to make this method named update

Reply:
will modify function name in follow-up PR